Source/Core/UICommon/NetPlayIndex.cpp

```diff
@@ -0,0 +1,321 @@
+// Copyright 2019 Dolphin Emulator Project
+// Licensed under GPLv2+
+// Refer to the license.txt file included.
+
+#include "UICommon/NetPlayIndex.h"
+
+#include <picojson/picojson.h>
+
+#include "Common/HttpRequest.h"
+#include "Common/Thread.h"
+#include "Common/Version.h"
+
+#include "Core/Config/NetplaySettings.h"
+
+NetPlayIndex::NetPlayIndex()
+{
+}
+
+NetPlayIndex::~NetPlayIndex()
+{
+  if (!m_secret.empty())
+    Remove();
+}
+
+static std::optional<picojson::value> ParseResponse(std::vector<u8> response)
+{
+  std::string response_string(reinterpret_cast<char*>(response.data()), response.size());
+
+  picojson::value json;
+
+  auto error = picojson::parse(json, response_string);
+
+  if (!error.empty())
+    return {};
+
+  return json;
+}
+
+std::optional<std::vector<NetPlaySession>>
+NetPlayIndex::List(const std::map<std::string, std::string>& filters)
+{
+  Common::HttpRequest request;
+
+  std::string list_url = Config::Get(Config::NETPLAY_INDEX_URL) + "/v0/list";
+
+  if (!filters.empty())
+  {
+    list_url += "?";
+    for (const auto& filter : filters)
+    {
+      list_url += filter.first + "=" + request.EscapeComponent(filter.second) + "&";
+    }
+    list_url = list_url.substr(0, list_url.size() - 1);
+  }
+
+  auto response = request.Get(list_url);
+
+  if (!response)
+  {
+    m_last_error = "NO_RESPONSE";
+    return {};
+  }
+
+  auto json = ParseResponse(response.value());
+
+  if (!json)
+  {
+    m_last_error = "BAD_JSON";
+    return {};
+  }
+
+  const auto& status = json->get("status");
+
+  if (status.to_str() != "OK")
+  {
+    m_last_error = status.to_str();
+    return {};
+  }
+
+  const auto& entries = json->get("sessions");
+
+  std::vector<NetPlaySession> sessions;
+
+  for (const auto& entry : entries.get<picojson::array>())
+  {
+    NetPlaySession session;
+
+    const auto& name = entry.get("name");
+    const auto& region = entry.get("region");
+    const auto& method = entry.get("method");
+    const auto& game_id = entry.get("game");
+    const auto& server_id = entry.get("server_id");
+    const auto& has_password = entry.get("password");
+    const auto& player_count = entry.get("player_count");
+    const auto& port = entry.get("port");
+    const auto& in_game = entry.get("in_game");
+
+    if (!name.is<std::string>() || !region.is<std::string>() || !method.is<std::string>() ||
+        !server_id.is<std::string>() || !game_id.is<std::string>() || !has_password.is<bool>() ||
+        !player_count.is<double>() || !port.is<double>() || !in_game.is<bool>())
+      continue;
+
+    session.name = name.to_str();
+    session.region = region.to_str();
+    session.game_id = game_id.to_str();
+    session.server_id = server_id.to_str();
+    session.method = method.to_str();
+    session.has_password = has_password.get<bool>();
+    session.player_count = static_cast<int>(player_count.get<double>());
+    session.port = static_cast<int>(port.get<double>());
```

Comment:
I wonder if we could/should define `PICOJSON_USE_INT64` and use that instead of the double. JSON only knows a number type, but it still feels weird checking for `double` when we want an `int`.

Reply:
Pretty low-priority IMHO

Reply:
Perhaps you should use std::lround instead of casting. Integer values traveling through a double scare me.